src/option_learning.py

```diff
@@ -212,3 +219,205 @@ def update_segment_from_option_spec(self, segment: Segment,
+            return True
+        # If we fall outside of the space where the preconditions hold, also
+        # terminate. The assumption is that the option should take us from one
+        # abstract state to another without hitting any others in between.
+        if not grounded_op.preconditions.issubset(abs_state):
+            return True
+        # Not yet done.
+        return False
+
+
+class _SimpleOptionLearner(_OptionLearnerBase):
+    """The option learner that learns options from fixed-length input vectors.
+
+    That is, the the input data only includes objects whose state
+    changes from the first state in the segment to the last state in the
+    segment. The input data does not include data from other objects in
+    the scene, where the number of other objects would differ across
+    different instiantiations of the env.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        # While learning the policy, we record the map from each segment to
+        # the option parameterization, so we don't need to recompute it in
+        # update_segment_from_option_spec.
+        self._segment_to_grounding: Dict[Segment, Tuple[Sequence[Object],
+                                                        Array]] = {}
+
+    def learn_option_specs(self, strips_ops: List[STRIPSOperator],
+                           datastores: List[Datastore]) -> List[OptionSpec]:
+        # In this paradigm, the option initiable and termination are determined
+        # from the operators, so the main thing that needs to be learned is the
+        # option policy.
+        option_specs: List[Tuple[ParameterizedOption, List[Variable]]] = []
+
+        assert len(strips_ops) == len(datastores)
+
+        for op, datastore in zip(strips_ops, datastores):
+            print(f"\nLearning option for NSRT {op.name}")
+
+            X_regressor: List[Array] = []
+            Y_regressor = []
+
+            # The superset of all objects whose states we may include in the
+            # option params is op.parameters. But we only want to include
+            # those objects that have state changes (in at least some data).
+            changing_parameter_set = self._get_changing_parameters(datastore)
+            # Just to avoid confusion, we will insist that the order of the
+            # changing parameters is consistent with the order of the original.
+            changing_parameters = sorted(changing_parameter_set,
+                                         key=op.parameters.index)
+            del changing_parameter_set  # not used after this
+            assert changing_parameters == [
+                v for v in op.parameters if v in changing_parameters
+            ]
+
+            for segment, sub in datastore:
+                inv_sub = {v: o for o, v in sub.items()}
+                all_objects_in_operator = [inv_sub[v] for v in op.parameters]
+
+                # First, determine the absolute goal vector for this segment.
```

Comment:
@ashayathalye this "absolute goal vector" and "relative goal vector" idea is new. I'm hoping it can be the best of both relative and absolute. see also the corresponding change in the option initiable and policy. let me know if it makes sense.